overlord/restart/restart_test.go

```diff
@@ -0,0 +1,143 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package restart_test
+
+import (
+	"testing"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/overlord/restart"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func TestRestart(t *testing.T) { TestingT(t) }
+
+type restartSuite struct{}
+
+var _ = Suite(&restartSuite{})
+
+type testHandler struct {
+	restartRequested   bool
+	rebootAsExpected   bool
+	rebootDidNotHappen bool
+}
+
+func (h *testHandler) HandleRestart(t restart.RestartType) {
+	h.restartRequested = true
+}
+
+func (h *testHandler) RebootAsExpected(*state.State) error {
+	h.rebootAsExpected = true
+	return nil
+}
+
+func (h *testHandler) RebootDidNotHappen(*state.State) error {
+	h.rebootDidNotHappen = true
+	return nil
+}
+
+func (s *restartSuite) TestRequestRestartDaemon(c *C) {
+	st := state.New(nil)
+
+	st.Lock()
+	defer st.Unlock()
+
+	// unitialized
+	ok, t := restart.Pending(st)
+	c.Check(ok, Equals, false)
+	c.Check(t, Equals, restart.RestartUnset)
+
+	h := &testHandler{}
+
+	err := restart.Init(st, "boot-id-1", h)
```

Comment:
(nitpick^2) I suppose we could test that errors from the handler are returned as expected but given that we test that the handler is caleld it probably does not add enough for the extra work.